### ISSUE TITLE
Improving modularisation by allowing inputs from the main workflow into subworkflows

### ISSUE THREAD

**Issue by irm-codebase** [snakemake/snakemake#2779]
**Is your feature request related to a problem? Please describe.**

Yes. Sort of.

Our team works with very complex data pipelines for energy model generation. Sort of like this (other pipelines have 80+ steps):
![image](https://github.com/snakemake/snakemake/assets/72193617/c130b60b-a026-4c6d-9b89-9e8ab61e5ff4)

We've noticed that students struggle quite a lot to understand these pipelines, primarily because they have too many components. Although we've tried to "trim" them to the minimum allowed size, in many cases it really is not possible to remove more pieces for clarity.

I got excited when I found out that Snakemake has module functionality... but it seems it is too limiting. Modules must be "standalone", without inputs from your parent pipeline. This is very limiting.

**Describe the solution you'd like**

In fields like embedded software engineering and electronics, complexity is normally "abstracted" by tucking it away in sub-modules, with clearly defined inputs and outputs. Kind of like this:
![image](https://github.com/snakemake/snakemake/assets/72193617/ec57b148-e4ef-47b3-88c9-597260a43c58)

Allowing us to run sub-workflows while defining their inputs in the main workflow (and "grouping" their steps in the main DAG) would be really welcomed. I think many fields would greatly benefit from such a feature.

**Basically, having an easy way to define "input" files into sub-workflows, and grouping DAG rules of said sub-workflows would be a god send!**

(if this is already possible, please correct me! It will make our lives easier!)

**Describe alternatives you've considered**

So far we have tried to "group" functionality in individual `.smk` files, but this requires a lot of importing, and the DAG is generally still quite hard to explain to students or new colleagues.

Honestly we've started to look at other tools (Dagster, for example), but they do not work as well as Snakemake when sending jobs to slurm.

**Comment by ASLeonard** [comment]
It is a fragile hack, but rather than importing submodules you can just make a rule in your top snakefile that runs the submodule directly. This would also solve your DAG visualiation, because the complex submodules only appear as the single rule “Create_transport_timeseries” etc (which then spawn all their own jobs when run). `include: <submodule>` can sometimes slow DAG building too much or have naming clashes, so each approach has its own drawbacks.

```
rule Create_transport_timeseries:
    input:
        <all input needed to start submodule>
    output:
       <any output you want from the entire submodule>
    shell:
        'snakemake -s <submodule> ...'
```

**Comment by irm-codebase** [comment]
> It is a fragile hack, but rather than importing submodules you can just make a rule in your top snakefile that runs the submodule directly. This would also solve your DAG visualiation, because the complex submodules only appear as the single rule “Create_transport_timeseries” etc (which then spawn all their own jobs when run). `include: <submodule>` can sometimes slow DAG building too much or have naming clashes, so each approach has its own drawbacks.
> 
> ```
> rule Create_transport_timeseries:
>     input:
>         <all input needed to start submodule>
>     output:
>        <any output you want from the entire submodule>
>     shell:
>         'snakemake -s <submodule> ...'
> ```

Thank you for the input. This might work! Testing shows some issues, though.
With the following structure:

```
project
- Snakemake
- modules/
    - light/
        - Snakemake
```

I run
```python
rule run_light_submodule:
    input:
        "output/global.txt"
    output:
        "modules/light/output/final.txt"
    shell:
        'snakemake -s modules/light/Snakefile -c 1'
```

Which returns

```shell
LockException:
Error: Directory cannot be locked. Please make sure that no other Snakemake process is trying to create the same files in the following directory:
/home/ivanruizmanuel/Documents/git/snakemake_tests
If you are sure that no other instances of snakemake are running on this directory, the remaining lock was likely caused by a kill signal or a power loss. It can be removed with the --unlock argument.
[Wed Mar 27 15:15:13 2024]
Error in rule run_light_submodule:
    jobid: 8
    input: output/global.txt
    output: modules/light/output/final.txt
    shell:
        snakemake -s modules/light/Snakefile -c 1
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: .snakemake/log/2024-03-27T151513.412171.snakemake.log
WorkflowError:
At least one job did not complete successfully.
```

It seems like the subfolder is locked by snakemake... any advice?

**Comment by ASLeonard** [comment]
You should pass `--nolock` to any "submodule" snakemake calls, as well as any config/profile args you would normally run them with

**Comment by irm-codebase** [comment]
In the end, I just went with the module functionality.
It's... not perfect, and does not abstract the DAG, but it's enough for now.
